source/blog/2015-10-31-rspec-compsable-matchers.md

```diff
@@ -0,0 +1,220 @@
+defines whether or not the passed in object belongs in the same category.
+For example, `Array === []` returns `true`. This is because an empty array is
+an instance of the array class. `Array === {}` returns false because a hash is
+not an array. For regexes, `===` returns true if the passed string matches the
+regex, and false if it does not.
+
+The `===` method is used in Ruby case statements to decide if a branch should
+be taken. For example
+
+```Ruby
+case 3
+when (1...18)
+  puts "hello"
+end
+```
+
+will print hello, because `(1...18) === 3` returns `true`.
+
+The new RSpec matchers make heavy use of `===` internally, if you'd like to get
+more of an understanding of `===`, I'd heavily suggest you pop open IRB and
+play around with it a little before moving on.
+
+## RSpec matchers
+
+In an RSpec expectation expression like `expect(object_1).to eq(object_2)`
+there are two primary objects at play, the **target** and the **matcher**. The
+**target** object is created by "expect" and exposes `to` (and `to_not`) which
+takes a matcher and is responsible for invoking `matches?` on it.
+
+It's possible to see this happening in an IRB session, e.g:
+
+```Ruby
+>> require "rspec/expectations"
+=> true
+>> extend(RSpec::Matchers)
+=> main
+>> expect(3).to Object.new
+NoMethodError: undefined method `matches?' for #<Object:0x007fa20d1f4e68>
+    from /Users/sam/.gem/ruby/2.1.5/gems/rspec-expectations-3.2.1/lib/rspec/expectations/handler.rb:50:in `block in handle_matcher'
+    from /Users/sam/.gem/ruby/2.1.5/gems/rspec-expectations-3.2.1/lib/rspec/expectations/handler.rb:27:in `with_matcher'
+    from /Users/sam/.gem/ruby/2.1.5/gems/rspec-expectations-3.2.1/lib/rspec/expectations/handler.rb:48:in `handle_matcher'
+    from /Users/sam/.gem/ruby/2.1.5/gems/rspec-expectations-3.2.1/lib/rspec/expectations/expectation_target.rb:54:in `to'
+    from (irb):4
+    from /Users/sam/.rubies/ruby-2.1.5/bin/irb:11:in `<main>'
+>>
+```
+
+In this session, we extend the RSpec DSL into the main object and then call
+`expect(3).to Object.new`. This blows up because our new object does not
+implement the `matches?` method.
+
+In RSpec expectations matchers implement `matches?` to provide assertions for
+your test suite, some of these are complex but they can also be quite simple.
+For example, here's the logic of `matches?` for the `eq` matcher:
+
+```ruby
+def matches?(actual)
+  actual == expected
+end
+```
+
+## Composable RSpec matchers
+
+A composable RSpec matcher is one which accepts other RSpec matchers for it's
+arguments. These are higher order matchers such as `include`, `start_with` and
+`match` (matchers like `be_predicate` or `be > x` are more primative and
+generally aren't composable). Let's take a look at the `match` matcher as a
+place to start.
+
+In RSpec 2.xx the match matcher was mostly used to match regexes, and its
+`matches?` definition looked thus:
+
+```ruby
+def matches?(expected, actual)
+  actual.match expected
+end
+```
```

Comment:
This code snippet also doesn't render correctly.  (I'll stop commenting on them now -- assume it's that way for all of them).